pkg/entitycache/entity_cache.go

```diff
@@ -82,21 +80,23 @@ type EntityID struct {
 }
 
 // NewEntity creates a new entity from ID and IP address from the tagger.
-func NewEntity(id EntityID, ipAddress, name string, services []string) *Entity {
-	return &Entity{
-		ID:         id,
-		IPAddress:  ipAddress,
+func NewEntity(id EntityID, ipAddress, name string, services []string) *entitycachev1.Entity {
+	return &entitycachev1.Entity{
+		EntityId: &entitycachev1.Entity_EntityID{
+			Prefix: id.Prefix,
+			Uid:    id.UID,
+		},
+		IpAddress:  ipAddress,
 		Services:   services,
 		EntityName: name,
 	}
 }
 
 // EntityCache maps IP addresses and Entity names to entities.
 type EntityCache struct {
-	entitycachev1.UnimplementedEntityCacheServiceServer
 	sync.RWMutex
-	entitiesByIP   map[string]*Entity
-	entitiesByName map[string]*Entity
+	entitiesByIP   *entitycachev1.Entities
```

Comment:
Track these maps as a EntityCache protobuf message that contains both maps. 

That message can be sent to the user on request. Make sure to do a deep copy when returning anything over grpc service methods as lock will be released while protobuf gets marshaled.

Reply:
Got it! - since the lock will be released when protobuf gets marshaled, using deep copy which does not share references can prevent data race.